File: api/_lib/mcp/tools/meals.ts
import type { McpToolDef } from '../protocol.js';
import { optionalBoolean, requireDateRange } from '../args.js';
import { fetchAllPages } from '../../pagination.js';
import type { MealRow } from '../../../../src/lib/db/types.js';
import { mealCalories, rowToMeal, sumDayMacros } from '../../../../src/lib/nutrition/mapping.js';

// Nutrition tool: meals in a range with per-day macro totals (calories are
// stored-or-derived via the same 4/4/9 Atwater rule the app displays).

const MAX_MEAL_DAYS = 31;

export const getMealsTool: McpToolDef = {
  name: 'get_meals',
  description:
    'Meals logged in a date range (max 31 days) with per-day macro totals (calories, protein, carbs, fat). ' +
    'Set include_items true for the individual meals behind each day.',
  inputSchema: {
    type: 'object',
    properties: {
      start_date: { type: 'string', description: 'Range start, YYYY-MM-DD (inclusive).' },
      end_date: { type: 'string', description: 'Range end, YYYY-MM-DD (inclusive).' },
      include_items: { type: 'boolean', description: 'Include individual meals per day (default false).' },
    },
    required: ['start_date', 'end_date'],
  },
  async run(supabase, userId, args) {
    const { startDate, endDate } = requireDateRange(args, MAX_MEAL_DAYS);
    const includeItems = optionalBoolean(args, 'include_items', false);

    const rows = await fetchAllPages<MealRow>('meals', (from, to) =>
      supabase
        .from('meals')
        .select('*')
        .eq('user_id', userId)
        .gte('date', startDate)
        .lte('date', endDate)
        .order('date', { ascending: true })
        .order('id', { ascending: true })
        .range(from, to),
    );

    const byDate = new Map<string, ReturnType<typeof rowToMeal>[]>();
    for (const row of rows) {
      const meal = rowToMeal(row);
      byDate.set(meal.date, [...(byDate.get(meal.date) ?? []), meal]);
    }

    const days = [...byDate.entries()]
      .sort((a, b) => a[0].localeCompare(b[0]))
      .map(([date, meals]) => ({
        date,
        meal_count: meals.length,
        totals: sumDayMacros(meals),
        meals: includeItems
          ? meals.map(m => ({
              title: m.title,
              time: m.time ?? null,
              meal_type: m.mealType ?? null,
              calories: mealCalories(m),
              protein_g: m.proteinG ?? null,
              carbs_g: m.carbsG ?? null,
              fiber_g: m.fiberG ?? null,
              sugar_g: m.sugarG ?? null,
              fat_total_g: m.fatTotalG ?? null,
              notes: m.notes || null,
            }))
          : undefined,
      }));

    return { start_date: startDate, end_date: endDate, days };
  },
};

File: ios/Packages/ApexCore/Sources/ApexCore/Models/Meals.swift
import Foundation

/// `POST /api/query { tool: "get_meals", args: { start_date, end_date, include_items } }`
/// (`api/_lib/mcp/tools/meals.ts`). The server sums the macros — including the
/// Atwater fallback for meals with no stored calories — so the app never adds
/// grams up itself (`src/lib/nutrition` has the tests).
public struct MealsQueryResult: Codable, Sendable, Equatable {
    public let startDate: String
    public let endDate: String
    public let days: [Day]

    enum CodingKeys: String, CodingKey {
        case startDate = "start_date"
        case endDate = "end_date"
        case days
    }

    public struct Day: Codable, Sendable, Equatable {
        public let date: String
        public let mealCount: Int
        public let totals: Totals
        /// Present only when the query asked for `include_items`.
        public let meals: [Item]?

        enum CodingKeys: String, CodingKey {
            case date
            case mealCount = "meal_count"
            case totals, meals
        }
    }

    /// Display-rounded by the server: whole calories, tenth-gram macros.
    public struct Totals: Codable, Sendable, Equatable {
        public let calories: Double
        public let proteinG: Double
        public let carbsG: Double
        public let fatTotalG: Double
    }

    /// One logged meal. Since W10 it carries its `id` (the phone edits and
    /// deletes through `/api/meals?id=`) and the whole stored fat split, so
    /// the composer can reopen it exactly as it was saved. Both are optional
    /// in the decoder so a response from before W10 still reads.
    public struct Item: Codable, Sendable, Equatable {
        public let id: String?
        public let title: String
        public let time: String?
        public let mealType: String?
        public let calories: Double?
        public let proteinG: Double?
        public let carbsG: Double?
        public let fiberG: Double?
        public let sugarG: Double?
        public let fatTotalG: Double?
        public let fatSaturatedG: Double?
        public let fatTransG: Double?
        public let alcoholG: Double?
        public let notes: String?

        enum CodingKeys: String, CodingKey {
            case id, title, time, calories, notes
            case mealType = "meal_type"
            case proteinG = "protein_g"
            case carbsG = "carbs_g"
            case fiberG = "fiber_g"
            case sugarG = "sugar_g"
            case fatTotalG = "fat_total_g"
            case fatSaturatedG = "fat_saturated_g"
            case fatTransG = "fat_trans_g"
            case alcoholG = "alcohol_g"
        }

        public init(
            id: String? = nil, title: String, time: String? = nil, mealType: String? = nil, calories: Double? = nil,
            proteinG: Double? = nil, carbsG: Double? = nil, fiberG: Double? = nil, sugarG: Double? = nil,
            fatTotalG: Double? = nil, fatSaturatedG: Double? = nil, fatTransG: Double? = nil, alcoholG: Double? = nil,
            notes: String? = nil
        ) {
            self.id = id
            self.title = title
            self.time = time
            self.mealType = mealType
            self.calories = calories
            self.proteinG = proteinG
            self.carbsG = carbsG
            self.fiberG = fiberG
            self.sugarG = sugarG
            self.fatTotalG = fatTotalG
            self.fatSaturatedG = fatSaturatedG
            self.fatTransG = fatTransG
            self.alcoholG = alcoholG
            self.notes = notes
        }
    }

    /// Arguments for the query, dated `YYYY-MM-DD`; the server caps a call at 31 days.
    public static func args(startDate: String, endDate: String, includeItems: Bool = true) -> [String: JSONValue] {
        [
            "start_date": .string(startDate),
            "end_date": .string(endDate),
            "include_items": .bool(includeItems),
        ]
    }
}

/// A meal template (`GET /api/meal-favorites`, W10): the same fields minus
/// the calendar placement, camelCase because the server maps the row
/// (`rowToFavorite`) before answering. Ids are `mealfav-<uuid>`, minted on
/// the phone the way the web mints them; a same-title save reuses the
/// existing id so "save again" overwrites rather than duplicates.
public struct MealFavorite: Codable, Sendable, Equatable, Identifiable {
    public let id: String
    public let title: String
    public let mealType: String?
    public let calories: Double?
    public let proteinG: Double?
    public let carbsG: Double?
    public let fiberG: Double?
    public let sugarG: Double?
    public let fatTotalG: Double?
    public let fatSaturatedG: Double?
    public let fatTransG: Double?
    public let alcoholG: Double?
    public let notes: String

    public init(
        id: String, title: String, mealType: String? = nil, calories: Double? = nil, proteinG: Double? = nil,
        carbsG: Double? = nil, fiberG: Double? = nil, sugarG: Double? = nil, fatTotalG: Double? = nil,
        fatSaturatedG: Double? = nil, fatTransG: Double? = nil, alcoholG: Double? = nil, notes: String = ""
    ) {
        self.id = id
        self.title = title
        self.mealType = mealType
        self.calories = calories
        self.proteinG = proteinG
        self.carbsG = carbsG
        self.fiberG = fiberG
        self.sugarG = sugarG
        self.fatTotalG = fatTotalG
        self.fatSaturatedG = fatSaturatedG
        self.fatTransG = fatTransG
        self.alcoholG = alcoholG
        self.notes = notes
    }
}

public struct MealFavoritesResponse: Codable, Sendable, Equatable {
    public let favorites: [MealFavorite]
}

/// The ids the web mints for meals and favorites (`MealsContext.tsx`).
public enum MealID {
    public static func mint() -> String { "meal-" + UUID().uuidString.lowercased() }
    public static func mintFavorite() -> String { "mealfav-" + UUID().uuidString.lowercased() }
}
